src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -127,56 +177,122 @@ def split_closures(node: ir.FencilDefinition) -> FencilWithTemporaries:
-        processed.
-        """
-        if isinstance(arg, ir.SymRef):
-            return arg
-        if (
-            isinstance(arg, ir.FunCall)
-            and isinstance(arg.fun, ir.FunCall)
-            and arg.fun.fun == ir.SymRef(id="lift")
-        ):
-            assert len(arg.fun.args) == 1
-            ref = ir.SymRef(id=f"_gtmp_{len(tmps)}")
-            tmps.append(ir.Sym(id=ref.id))
-            unlifted = ir.FunCall(fun=arg.fun.args[0], args=arg.args)
-            stencil_stack.append((ref, unlifted))
-            return ref
-        raise AssertionError()
+    uid_gen_tmps = UIDGenerator(prefix="_tmp")
+
+    type_inference.infer_all(node, offset_provider=offset_provider, save_to_annex=True)
 
-    closures = []
+    tmps: list[ir.Sym] = []
+
+    closures: list[ir.StencilClosure] = []
     for closure in reversed(node.closures):
-        wrapped_stencil = ir.FunCall(fun=closure.stencil, args=closure.inputs)
-        popped_stencil = PopupTmps().visit(wrapped_stencil)
+        closure_stack: list[ir.StencilClosure] = [closure]
+        while closure_stack:
+            current_closure: ir.StencilClosure = closure_stack.pop()
+
+            if current_closure.stencil == im.ref("deref"):
+                closures.append(current_closure)
+                continue
+
+            # TODO(tehrengruber): extracting from a scan is untested.
+            is_scan: bool = isinstance(
+                current_closure.stencil, ir.FunCall
+            ) and current_closure.stencil.fun == im.ref("scan")
+            current_closure_stencil = (
+                current_closure.stencil if not is_scan else current_closure.stencil.args[0]  # type: ignore[attr-defined]  # ensured by is_scan
+            )
 
-        stencil_stack = [(closure.output, popped_stencil)]
-        domain = closure.domain
-        while stencil_stack:
-            output, call = stencil_stack.pop()
-            closure = ir.StencilClosure(
-                domain=domain,
-                stencil=call.fun,
-                output=output,
-                inputs=[handle_arg(arg) for arg in call.args],
+            stencil_body, extracted_lifts, _ = extract_subexpression(
+                current_closure_stencil.expr,
+                temporary_extraction_predicate,
+                uid_gen_tmps,
+                once_only=True,
+                deepest_expr_first=True,
             )
-            closures.append(closure)
-            domain = AUTO_DOMAIN
+
+            if extracted_lifts:
+                for tmp_sym, lift_expr in extracted_lifts.items():
+                    # make sure the applied lift is not capturing anything except of closure params
+                    used_symbol_refs = collect_symbol_refs(lift_expr)
+                    assert not (
+                        set(used_symbol_refs)
+                        - {param.id for param in current_closure_stencil.params}
+                    )
```

Comment:
nit: I'd prefer something like `assert _applied_lift_captures_only_closure_params(lift_expr, current_closure_stencil.params)` because I don't like comments that could be code.

Reply:
I don't see that point here to be honest. The comment can rod the same way the name of the function can. If I change the code without updating the comment that appears as likely as changing a function without updating its name (otherwise one is discriminating comments :-P). Your choice, but the very long function name in comparison to its content seems a little odd.

Reply:
Conclusion: make function :-P